NeoMathEngine/include/NeoMathEngine/ThreadPool.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright � 2017-2022 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <NeoMathEngine/NeoMathEngineDefs.h>
+#include <NeoMathEngine/CrtAllocatedObject.h>
+#include <NeoMathEngine/NeoMathEngineException.h>
+
+namespace NeoML {
+
+// The class provides thread pool functionality.
+class NEOMATHENGINE_API IThreadPool : public CCrtAllocatedObject {
+public:
+	// Interface for pool task.
+	typedef void(*TFunction)(int threadIndex, void* params);
+
+	virtual ~IThreadPool();
+
+	// Returns the number of threads in the pool.
+	virtual int Size() const = 0;
+
+	// Adds a task with parameters for the given thread.
+	virtual bool AddTask( int threadIndex, TFunction function, void* params ) = 0;
+
+	// Waits for all tasks to complete.
+	virtual void WaitAllTask() = 0;
+
+	// Stops all threads and waits for them to complete.
+	virtual void StopAndWait() = 0;
```

Comment:
`StopAndWait()` can be removed from `IThreadPool` interface.

Reply:
1) This function calls from only once, so it can inline directly onto the called side.
2) This functions calls from ThreadPool destructor, so it must be non-virtual.